# Laboratorio 5: Clasificación 🤗

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2024</strong></center>


### Cuerpo Docente:

- Profesores: Stefano Schiappacasse, Sebastián Tinoco
- Auxiliares: Melanie Peña, Valentina Rojas
- Ayudantes: Angelo Muñoz, Valentina Zúñiga

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Sebastián Acuña U.
- Nombre de alumno 2: Martín Guzmán S.

### **Link de repositorio de GitHub:** [Repositorio](https://github.com/sebastianacunau/MDS7202-Laboratorios-y-Proyecto)

### Temas a tratar
- Clasificación en problemas desbalanceados
- Lightgbm y xgboost
- Pipelines

## Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 6 días de plazo con descuento de 1 punto por día. Entregas Martes a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria, pero se recomienda fuertemente asistir.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.




### Objetivos principales del laboratorio
- Comprender cómo trabajar con problemas de clasificación con clases desbalanceadas.
- Aplicar los modelos lightgbm y xgboost.
- Practicar Pipelines

# Parte Teórica [12 puntos]




1. Explique cuál es la diferencia entre los datos de entrenamiento y validación. [1 punto]

2. Explique cuál es el principal desafío al trabajar problemas de clasificación con data no supervisada. [1 punto]

3. Explique en **sus palabras** qué es la matriz de confusión y para qué se utiliza. [1 puntos]

4. Escriba la fórmula de las siguientes métricas y explique con **sus palabras** cómo se interpretan. [1 punto cada uno]

  * Accuracy
  * Precision
  * Recall
  * F1 score

5. Explique qué métrica recomendaría para los siguientes contextos de clasificación. [1 punto cada uno]

  * Mantenimiento predictivo de fallas de maquinaria pesada en la industria minera.  
  * Detección de enfermedades altamente contagiosas.
  * Aprobación de créditos de alto riesgo.
  * Detección de crímenes.

6. Explique qué es la calibración de modelos y para qué se usa. [1 punto]


**Respuesta:**

1. La diferencia es que los datos de entrenamiento, tal como su nombre indica, se ocupan para entrenar al modelo seleccionado, con el objetivo de estimar parámetros o generar criterios de clasificación. Por otro lado, los datos de validación sirven para evaluar la calidad de ajuste del modelo ya entrenado, calculando métricas en la fase de evaluación que permiten determinar qué tan bien se comporta el modelo con datos que no conoce.

2. El principal desafio al trabajar en problemas de clasificación con data no supervisada es que no hay una etiqueta real en la data para poder comparar el output de nuestro modelo con la data real, lo que dificulta la interpretabilidad de los modelos y determinar cuántas clases hay realmente en el problema, principalmente porque dependen mucho del modelo y de las metricas de performance escogidas.

3. La matriz de confusión es una matriz cuadrada que compara las clases predichas con las verdaderas, en el contexto de modelos de clasificación con aprendizaje supervisado. Si los valores a predecir se clasifican en $n$ clases, entonces la matriz es de tamaño $n \times n$, tal que en un eje se encuentran las clases reales de las instancias (usualmente el vertical, correspondiente a las filas), y en el otro eje se encuentran las predicciones. Así, la componente $(i,j)$ corresponde a la cantidad de datos que pertenecen a la clase $i$ pero la predicción fue $j$. Es decir, las predicciones correctas del modelo se encuentran en la diagonal, mientras que todo el resto de las casillas corresponde a errores de predicción.

4. En el contexto de la clasificación con aprendizaje supervisado, interesa conocer qué tan acertadas son las predicciones, realizando comparaciones entre los datos correctamente predichos y el total. De allí nacen las siguientes métricas para evaluar la calidad de la predicción de un modelo. Primero definimos la siguiente notación (supondremos que existen dos categorías, positivo y negativo, pero con suficiente cuidado es generalizable a $n$):

    * $TP$: La cantidad de datos que fueron correctamente asignados como positivos (true positives).
    * $TN$: La cantidad de datos que fueron correctamente asignados como negativos (true negatives).
    * $FP$: La cantidad de datos que fueron incorrectamente asignados como positivos (false positives).
    * $FN$: La cantidad de datos que fueron incorrectamente asignados como negativos (false negatives).

Con ello, se definen las siguientes métricas y sus respectivas fórmulas:

* Accuracy: Mide la proporción de los datos correctamente etiquetados sobre el total.

$$
Accuracy = \frac{TP+TN}{TP+TN+FP+FN}
$$

* Precision: De todos los datos catalogados como positivos, mide la proporción de ellos que efectivamente están bien etiquetados.

$$
Precision = \frac{TP}{TP+FP}
$$

* Recall: Mide la proporción de datos correctamente catalogados como positivos sobre todos los datos realmente positivos.

$$
Recall = \frac{TP}{TP+FN}
$$

* F1 score: Corresponde a la media armónica entre las métricas precision y recall. Es una métrica más balanceada cuando no se prioriza la minimización de un error a costa del otro.

$$
F_1 = 2 \cdot \frac{Precision \cdot Recall}{Precision + Recall} = \frac{2TP}{2TP+FP+FN}
$$

5. Las métricas recomendadas a optimizar para cada caso son:

    * Mantenimiento predictivo de fallas de maquinaria pesada en la industria minera: Depende de qué es más costoso, una detención por falla o una serie de detenciones preventivas a cambio de prevenir fallas de maquinaria. Suponiendo que es lo primero entonces conviene optimizar Recall, pues lo ideal es identificar todos las fallas posibles, aún si a veces se predicen algunos falsos positivos. Dicho de otro modo, el objetivo principal es minimizar los falsos negativos.
    * Detección de enfermedades altamente contagiosas: Recall. Al igual que en el caso anterior, lo que se busca es minimizar los falsos negativos, aún si eso genera un leve incremento en la cantidad de falsos positivos.
    * Aprobación de créditos de alto riesgo: Precision. No es tan grave dejar sin un crédito de alto riesgo a alguien calificado, pero el banco no se puede permitir que una persona riesgosa reciba uno (falso positivo). En ese sentido, la precisión es una métrica un poco más conservadora. 
    * Detección de crímenes: F1 score. Un gobierno o la policía local de un país no pueden permitirse encarcelar a alguien por un crimen que no cometió, pero también es muy necesario detectar la mayor cantidad de criminales, de modo que el F1 score provee ese balance.

6. La calibración de modelos es el proceso mediante el cual verificamos si las probabilidades que entrega un modelo de clasificación realmente reflejan la realidad. En otras palabras, queremos saber si cuando el modelo dice, por ejemplo, que hay un 80% de probabilidad de que algo ocurra, eso significa que en efecto, en promedio, eso ocurre el 80% de las veces.
    
    Esto es especialmente importante en modelos que entregan probabilidades como salida, porque no basta con que el modelo acierte o se equivoque en su clasificación: también importa qué tan confiado está y si esa confianza está justificada.

    Un modelo está bien calibrado si, por ejemplo, de todos los casos en los que asigna un 70% de probabilidad a una clase, aproximadamente el 70% de esos casos realmente pertenecen a esa clase. Si asigna probabilidades muy altas pero se equivoca frecuentemente, o al revés, sugiere estar poco seguro pero acierta mucho, entonces el modelo está mal calibrado.

# Parte práctica [48 puntos]

<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1BnO4tyh3vM2P199Ec9s3JjngQ4qQ9seP"
" width="300">
</p>


Tras el trágico despido de la mítica mascota de Maipú, Renacín decide adentrarse como consultor en el mercado futbolero, el cuál (para variar...) está cargado en especulaciones.

Como su principal tarea será asesorar a los directivos de los clubes sobre cuál jugador comprar y cuál no, Renacín desea generar modelos predictivos que evaluén distintas características de los jugadores; todo con el fin de tomar decisiones concretas basadas en los datos.

Sin embargo, su condición de corpóreo le impidió tomar la versión anterior de MDS7202, por lo que Renacín contrata a su equipo para lograr su objetivo final. Dado que aún tiene fuertes vínculos con la dirección de deportes de la municipalidad, el corpóreo le entrega base de datos con las estadísticas de cada jugador para que su equipo empieze a trabajar ya con un dataset listo para ser usado.


**Los Datos**

Para este laboratorio deberán trabajar con el csv `statsplayers.csv`, donde deberán aplicar algoritmos de aprendizaje supervisado de clasificación en base a características que describen de jugadores de fútbol.

Para comenzar cargue el dataset señalado y a continuación vea el reporte **`Player_Stats_Report.html`** (adjunto en la carpeta del enunciado) que describe las características principales del `DataFrame`.

In [ ]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = 'Dirección donde tiene los archivos en el Drive'
except:
    print('Ignorando conexión drive-colab')

In [20]:
# Librerías a utilizar
import pandas as pd
import numpy as np
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler, LabelEncoder
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, recall_score, accuracy_score, precision_score
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier

In [21]:
data = pd.read_csv('stats_players.csv')
data.head()

Name Nationality National_Position Club_Position  Height  \
0  Cristiano Ronaldo    Portugal                LS            LW     185   
1       Lionel Messi   Argentina                RW            RW     170   
2             Neymar      Brazil                LW            LW     174   
3        Luis Suárez     Uruguay                LS            ST     182   
4       Manuel Neuer     Germany                GK            GK     193   

   Weight Preffered_Foot  Age        Work_Rate  Weak_foot  ...  Agility  \
0      80          Right   32       High / Low          4  ...       90   
1      72           Left   29  Medium / Medium          4  ...       90   
2      68          Right   25    High / Medium          5  ...       96   
3      85          Right   30    High / Medium          4  ...       86   
4      92          Right   31  Medium / Medium          4  ...       52   

   Jumping  Heading  Shot_Power  Finishing  Long_Shots  Curve  \
0       95       85          92         93          90     81   
1       68       71          85         95          88     89   
2       61       62          78         89          77     79   
3       69       77          87         94          86     86   
4       78       25          25         13          16     14   

   Freekick_Accuracy  Penalties  Volleys  
0                 76         85       88  
1                 90         74       85  
2                 84         81       83  
3                 84         85       88  
4                 11         47       11  

[5 rows x 39 columns]

## 1. Predicción de Seleccionados Nacionales [14 puntos]

<p align="center">
  <img src="https://www.futuro.cl/wp-content/uploads/2016/06/chile-argentina-meme-12.jpg" width="300">
</p>



### 1.1 Preprocesamiento [5 puntos]

Tareas:

1. Genere los labels para la clasificación binaria en una variable llamada `label`. Para esto, trabaje sobre el atributo `National_Position` suponiendo que los valores nulos son jugadores no seleccionados para representar a su país. [Sin puntaje]

2. Hecho esto, ¿cuántos ejemplos se tienen por cada clase? Comente lo que observa. [1 punto]

3. Genere un `ColumnTransformer` en donde especifique las transformaciones que hay que realizar para cada columna (por ejemplo StandardScaler, MinMaxScaler, OneHotEncoder, etc...) para que puedan ser utilizadas correctamente por el modelo predictivo y guárdelo una variable llamada `col_transformer`. [2 puntos]

4. Comente y justifique las transformaciones elegidas sobre cada una de las variables (para esto utilice el material `Player_Stats_Report.html` que viene en el zip del lab), al igual que las transformaciones aplicadas. [2 puntos]

**Respuesta:**

In [22]:
data['label'] = data['National_Position'].notnull().astype(int)
data['label'].value_counts()

label
0    16513
1     1075
Name: count, dtype: int64

Del total de jugadores, hay 1075 que son seleccionados nacionales, lo que corresponde a aproximadamente el 6.1% de los datos. Esto indica que si queremos predecir los jugadores que son seleccionados nacionales, debemos tener un alto nivel en las métricas para que el modelo se pueda considerar relevante.

In [31]:
numerical = data.select_dtypes(include=['int']).columns.tolist()
categorical = data.select_dtypes(include=['object']).columns.tolist()

numerical.remove('label')

In [32]:
# Define imputadores para valores faltantes
cat_imputer = SimpleImputer(strategy='most_frequent')  
num_imputer = SimpleImputer(strategy='mean')          

# Define el ColumnTransformer con imputación y transformación
col_transformer = ColumnTransformer(
    transformers=[
        ('cat', Pipeline([
            ('imputer', cat_imputer),  # Imputar valores faltantes
            ('encoder', OneHotEncoder(handle_unknown='ignore'))  # Codificar categóricas
        ]), categorical),
        ('num', Pipeline([
            ('imputer', num_imputer),  # Imputar valores faltantes
            ('scaler', StandardScaler())  # Escalar numéricas
        ]), numerical)])

### 1.2 Entrenamiento [3 puntos]

Ahora, vamos a entrenar los pipelines generados en los pasos anteriores. Para esto, debe realizar las siguientes tareas:

1. Separe los datos de entrenamiento en un conjunto de entrenamiento y de prueba  (la proporción queda a su juicio). En este paso, seleccione los ejemplos de forma aleatoria e intente mantener la distribución original de labels de cada clase en los conjuntos de prueba/entrenamiento. (vea la documentación de `train_test_split`). [1 puntos]


2. Defina un pipeline llamado `pipeline_xgboost` y otro llamado `pipeline_lightgbm`. Estos pipelines deben tener el mismo ColumnTransformer definido en la sección de preprocesamiento, pero deben variar los clasificadores de acuerdo al nombre de cada pipeline. [1 puntos]

3. Entrene los pipelines. [1 punto]



**Respuesta:**

In [33]:
x = data.drop(columns=['label'])
y = data['label']
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42, stratify=y)


pipeline_xgboost = Pipeline(
    steps=[
        ('Preprocessor', col_transformer),
        ('Classifier', XGBClassifier(random_state=42))
    ]
)

pipeline_lightgbm = Pipeline(
    steps=[
        ('Preprocessor', col_transformer),
        ('Classifier', LGBMClassifier(random_state=42))
    ]
)


pipeline_xgboost.fit(x_train, y_train)
pipeline_lightgbm.fit(x_train, y_train)

Pipeline(steps=[('Preprocessor',
                 ColumnTransformer(transformers=[('cat',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('encoder',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['Name', 'Nationality',
                                                   'National_Position',
                                                   'Club_Position',
                                                   'Preffered_Foot',
                                                   'Work_Rate']),
                                                 ('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   Standard...
                                                   'Ball_Control', 'Dribbling',
                                                   'Marking', 'Sliding_Tackle',
                                                   'Standing_Tackle',
                                                   'Aggression', 'Reactions',
                                                   'Interceptions', 'Vision',
                                                   'Composure', 'Crossing',
                                                   'Short_Pass', 'Long_Pass',
                                                   'Acceleration', 'Speed',
                                                   'Stamina', 'Strength',
                                                   'Balance', 'Agility',
                                                   'Jumping', 'Heading',
                                                   'Shot_Power', 'Finishing',
                                                   'Long_Shots', 'Curve', ...])])),
                ('Classifier', LGBMClassifier(random_state=42))])

### 1.3 Resultados [6 puntos]

1. Calcule las métricas accuracy, precisión y recall de la clase positiva (la que indica que un jugador es seleccionado nacional) para evaluar el rendimiento de los distintos modelos. Verifique sus resultados usando `classification_report`. [2 puntos]

2. Explique qué implican los valores de accuracy, precisión y recall de la clase positiva (la que indica que un jugador es seleccionado nacional) y cómo influye la cantidad de ejemplos por clase en los resultados obtenidos. [2 puntos]

3. Explique qué métrica le parece más adecuada y concluya qué modelo tiene un mejor desempeño. [2 puntos]

**Respuesta:**

2.⁠ ⁠Al momento de analizar las métricas de evaluación en la clase positiva. Se tiene un accuracy de 0.97 en ambos modelos, lo cual considera que los resultados reales son el 97% del total de los resultados. En la lectura de la métrica recall, se tiene un 0.61 en XGboost y 0.57 en LightGBM, es decir, el ratio entre los resultados positivos verdaderos y la suma entre positivo verdadero y negativo falso conllevan un porcentaje un poco mejor que un modelo aleatorio. Finalmente, se tiene una precisión e 0.91 en ambos modelos, lo que significa que el ratio entre los positivos verdaderos y la suma entre los positivos verdaderos y los falsos positivos es bastante alta. 


3.⁠ ⁠Ya que tenemos resultados muy parecidos en las métricas de accuracy y precisión, evaluamos por la métrica recall, con lo que decidiríamos por el modelo de XGBoost ya que tiene un ratio de más alto entre positivos verdaderos y los resultados predichos.

In [34]:
y_pred_xgboost = pipeline_xgboost.predict(x_test)
y_pred_lightgbm = pipeline_lightgbm.predict(x_test)

classification_report_xgboost = classification_report(y_test, y_pred_xgboost)
classification_report_lightgbm = classification_report(y_test, y_pred_lightgbm)
accuracy_xgboost = accuracy_score(y_test,y_pred_xgboost)    
accuracy_lightgbm = accuracy_score(y_test,y_pred_lightgbm)


print('XGBoost Classification Report') 
print(classification_report_xgboost)
print('XGBoost Accuracy:', accuracy_xgboost)


print('LightGBM Classification Report') 
print(classification_report_lightgbm)    
print('LightGBM Accuracy:', accuracy_lightgbm)

AttributeError: 'super' object has no attribute '__sklearn_tags__'

## 2. Predicción de posiciones de jugadores [4 puntos]

<p align="center">
  <img src="https://pbs.twimg.com/media/E1rfA1aWEAYU6Ny.jpg" width="300">
</p>

En una nueva jornada de desmesuradas transacciones deportivas, Renacín escuchó a sus colegas discutir acerca de que el precio de cada jugador depende en gran medida de la posición en la cancha en la que juega. Y además, que hay bastantes jugadores nuevos que no tienen muy claro en que posición verdaderamente brillarían, por lo que actualmente puede que actualmente estén jugando en posiciones sub-optimas.

Viendo que los resultados del primer análisis no son tan esperanzadores, el corporeo los comanda a cambiar su tarea: ahora, les solicita que construyan un clasificador enfocado en predecir la mejor posición de los jugadores en la cancha según sus características.

Para lograr esto, primero, les pide que etiqueten de la siguiente manera los valores que aparecen en el atributo `Club_Position`, pidiendo que agrupen los valores en los siguientes grupos:

**Nota**:  Renacín les recalca que **no deben utilizar los valores ```Sub``` y ```Res``` de esta columna**.

```python
ataque = ['ST', 'CF']
central_ataque = ['RW', 'CAM', 'LW']
central = ['RM', 'CM', 'LM']
central_defensa = ['RWB', 'CDM', 'LWB']
defensa = ['RB', 'CB', 'LB']
arquero = ['GK']
```

La elección del clasificador se justificar en base a la siguiente [guía](https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html) y se deben comentar los resultados obtenidos en la clasificación.

**Tareas:** [1 punto por tarea]

1. En un nuevo dataframe, aplique las etiquetas descritas anteriormente en cada uno de los valores señalados en esta sección y guárdelos en la variable `label`.
2. Cuente cuántos por clase quedan.
3. Entrene el nuevo pipeline y ejecute una evaluación de este.  
4. Comente los resultados obtenidos.

**Respuesta:**

In [27]:
df = pd.read_csv('stats_players.csv')

ataque = ['ST', 'CF']
central_ataque = ['RW', 'CAM', 'LW']
central = ['RM', 'CM', 'LM']
central_defensa = ['RWB', 'CDM', 'LWB']
defensa = ['RB', 'CB', 'LB']
arquero = ['GK']

def label_position(position):
    if position in ataque:
        return 'Ataque'
    elif position in central_ataque:
        return 'Central Ataque'
    elif position in central:
        return 'Central'
    elif position in central_defensa:
        return 'Central Defensa'
    elif position in defensa:
        return 'Defensa'
    elif position in arquero:
        return 'Arquero'
     

df['label'] = df['National_Position'].apply(label_position)
df = df[df['label'].notnull()]
df['label'].value_counts()  

numerical1 = df.select_dtypes(include=['int']).columns.tolist()
categorical1 = df.select_dtypes(include=['object']).columns.tolist()
categorical1.remove('label')

In [28]:
# Define imputadores para valores faltantes
cat_imputer = SimpleImputer(strategy='most_frequent')  
num_imputer = SimpleImputer(strategy='mean')          

# Define el ColumnTransformer con imputación y transformación
col_transformer = ColumnTransformer(
    transformers=[
        ('cat', Pipeline([
            ('imputer', cat_imputer),  # Imputar valores faltantes
            ('encoder', OneHotEncoder(handle_unknown='ignore'))  # Codificar categóricas
        ]), categorical1),
        ('num', Pipeline([
            ('imputer', num_imputer),  # Imputar valores faltantes
            ('scaler', StandardScaler())  # Escalar numéricas
        ]), numerical1)])

In [29]:
label_encoder = LabelEncoder()
df['label_encoded'] = label_encoder.fit_transform(df['label'])


x1 = df.drop(columns=['label'])
y1 = df['label_encoded']
x_train1, x_test1, y_train1, y_test1 = train_test_split(x1, y1, test_size=0.2, random_state=42, stratify=y1)


pipeline_xgboost1 = Pipeline(
    steps=[
        ('Preprocessor', col_transformer),
        ('Classifier', XGBClassifier(random_state=42, verbose=0))
    ]
)

pipeline_lightgbm1 = Pipeline(
    steps=[
        ('Preprocessor', col_transformer),
        ('Classifier', LGBMClassifier(random_state=42, verbose=0))
    ]
)

pipeline_xgboost1.fit(x_train1, y_train1)
pipeline_lightgbm1.fit(x_train1, y_train1)


y_pred_xgboost1 = pipeline_xgboost1.predict(x_test1)
y_pred_lightgbm1 = pipeline_lightgbm1.predict(x_test1)

classification_report_xgboost1 = classification_report(y_test1, y_pred_xgboost1)
classification_report_lightgbm1 = classification_report(y_test1, y_pred_lightgbm1)
accuracy_xgboost1 = accuracy_score(y_test1,y_pred_xgboost1)
accuracy_lightgbm1 = accuracy_score(y_test1,y_pred_lightgbm1)

/Users/sebastianacunaurzua/opt/anaconda3/lib/python3.12/site-packages/xgboost/core.py:158: UserWarning: [17:27:31] WARNING: /var/folders/c_/qfmhj66j0tn016nkx_th4hxm0000gp/T/abs_b6qk1lz_ug/croot/xgboost-split_1724073748391/work/src/learner.cc:740: 
Parameters: { "verbose" } are not used.

  warnings.warn(smsg, UserWarning)


[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No f

AttributeError: 'super' object has no attribute '__sklearn_tags__'

In [30]:
print('XGBoost Classification Report')
print(classification_report_xgboost1)
print('XGBoost Accuracy:', accuracy_xgboost1)

print('LightGBM Classification Report')
print(classification_report_lightgbm1)
print('LightGBM Accuracy:', accuracy_lightgbm1)

XGBoost Classification Report


NameError: name 'classification_report_xgboost1' is not defined

## 3. Predicciones de Seleccionados Nacionales para el Jere Klein [30 puntos]

<center>
<img src='https://www.radioactiva.cl/wp-content/uploads/2024/04/Jere-Klein-1-768x432.webp' width=500 />

Después de alcanzar la fama como cantante urbano, Jere Klein decide explorar una nueva faceta. Con su amor por el fútbol y convencido de que los artistas urbanos poseen un talento y versatilidad excepcionales, Jere se embarca en un proyecto innovador: desarrollar un sistema de inteligencia artificial capaz de identificar a jugadores que tienen potencial para convertirse en futbolistas profesionales. Su teoría es que muchos artistas del género urbano chileno, con sus habilidades únicas y su disciplina, podrían destacarse también en el deporte. Con este sistema, Jere espera no solo abrir nuevas oportunidades para sus colegas artistas, sino también demostrar la amplia gama de talentos que pueden ofrecer.

### 3.1 ¿Qué modelo de árbol es más de "pana"? [10 puntos]

<center>
<img src='https://64.media.tumblr.com/39189215a7d3d96823cb359f35b44e05/tumblr_psmrhrR3Xw1qf5hjqo4_540.gif' width=300 />


**Tareas**


1. Considerando el la variable llamada `label` creada en la sección 1.1. Para determinar cuál modelo de árbol sería más adecuado para la tarea en cuestión, utilice PyCaret. Este deberá centrarse exclusivamente en modelos de tipo árbol. Jere ha especificado que busca un modelo que tome decisiones rápidamente y que tenga una baja tasa de falsos positivos, ya que planea invertir en estos jugadores. [3 puntos]

Para la comparación, utilice los siguientes modelos:

```python
['et', 'rf', 'dt', 'xgboost', 'lightgbm', 'catboost']
```

2. Explique en brevemente que son los modelos de la siguiente lista `['et', 'rf', 'dt']` y como funcionan. [3 punto]

3. Tras realizar la comparación de modelos, seleccione aquel que muestre el mejor rendimiento en términos de velocidad y precisión, especialmente en la reducción de falsos positivos. Utilice la función `evaluate_model` de PyCaret para revisar y analizar los resultados obtenidos en los siguientes aspectos:

  - **Confusión Matrix**: ¿Cómo se encuentran la tasa de verdaderos positivos y verdaderos negativos?
  - **Threshold**: ¿Es acaso el umbral por defecto del modelo el mejor para las predicciones?
  - **Feature Importance**: ¿Cuáles son las variables con mejor desempeño? ¿A qué podría deberse esto?
  - **Learning Curve**: ¿El modelo presenta algún problema?

  [4 puntos]

**Respuesta**

In [36]:
from pycaret.datasets import get_data
from pycaret.classification import *
import os

os.environ["PYCARET_CUSTOM_LOGGING_LEVEL"] = "CRITICAL"

# Configuración de PyCaret con los datos de entrenamiento y la variable objetivo
clf1 = setup(
    data=x_train,
    target='label_encoded',
    session_id=42,
    silent=True,
    use_gpu=True,
    fold=5,
    log_experiment=False,
    experiment_name='tree_models_comparison',
    html=False
)

# Comparar modelos de tipo árbol
tree_models = ['et', 'rf', 'dt', 'xgboost', 'lightgbm', 'catboost']
best_model = compare_models(include=tree_models, sort='Precision')

# Evaluar el mejor modelo
evaluate_model(best_model)

ModuleNotFoundError: No module named 'pycaret'

### 3.2 Reducción de dimensionalidad [14 puntos]

<center>
<img src='https://i.kym-cdn.com/photos/images/original/002/258/560/668.gif' width=400 />

A pesar de los resultados obtenidos previamente, el manager de Jere ha solicitado el entrenamiento de un modelo de XGBoost utilizando los datos disponibles. Además, se debe proceder a realizar una reducción de dimensionalidad basada en la importancia de las características.

Para llevar a cabo esta tarea:

1. Inicie entrenando un modelo XGBoost con todas las características disponibles. [2 puntos]

2. Una vez el modelo esté entrenado, evalúe y clasifique las características según su importancia de forma descendente. [2 puntos]

3. Utilice esta clasificación para ejecutar una búsqueda recursiva de eliminación de características, eliminando progresivamente las menos importantes y evaluando el impacto en el desempeño del modelo hasta identificar las N características más críticas. [2 puntos]

4. Con este conjunto reducido de características, entrene un nuevo modelo y evalúe su rendimiento. [2 puntos]

5. Posteriormente, responda a las siguientes preguntas para una comprensión más profunda de los cambios y beneficios:

  - ¿El rendimiento del modelo con las características seleccionadas es similar al del modelo original? ¿Cómo se comparan en términos de precisión y robustez? [2 puntos]
  - ¿Cuáles son los beneficios potenciales de eliminar variables del modelo? Considere factores como la simplificación del modelo, reducción del tiempo de entrenamiento, y mejora en la capacidad de generalización. [2 puntos]
  - Comente si el modelo con menor dimensionalidad es más sencillo de explicar. Explique brevemente por qué la eliminación de ciertas características puede facilitar la comprensión y la explicación del comportamiento del modelo. [2 puntos]

Notar que con esta metodologia buscamos encontrar un punto entermedio entre número de festures y desempeño. por esto, si observa que al aumentar festires el aumento es despreciable, puede no considerar agregar más features a su modelo.

**Respuesta**

In [ ]:
from xgboost import XGBClassifier
from sklearn.feature_selection import RFECV
from sklearn.metrics import classification_report, accuracy_score

# Entrenar un modelo XGBoost con todas las características disponibles
xgb_model_full = XGBClassifier(random_state=42)
xgb_model_full.fit(x_train, y_train)

# Evaluar y clasificar las características según su importancia
feature_importances = xgb_model_full.feature_importances_
sorted_features = sorted(zip(x_train.columns, feature_importances), key=lambda x: x[1], reverse=True)

print("Características ordenadas por importancia:")
for feature, importance in sorted_features:
    print(f"{feature}: {importance:.4f}")

# Búsqueda recursiva de eliminación de características (RFE)
rfe_selector = RFECV(estimator=xgb_model_full, step=1, cv=5, scoring='accuracy')
rfe_selector.fit(x_train, y_train)

# Seleccionar las características más importantes
selected_features = x_train.columns[rfe_selector.support_]
print("\nCaracterísticas seleccionadas:")
print(selected_features)

# Entrenar un nuevo modelo con el conjunto reducido de características
x_train_reduced = x_train[selected_features]
x_test_reduced = x_test[selected_features]

xgb_model_reduced = XGBClassifier(random_state=42)
xgb_model_reduced.fit(x_train_reduced, y_train)

# Evaluar el modelo reducido
y_pred_reduced = xgb_model_reduced.predict(x_test_reduced)
print("\nReporte de clasificación para el modelo reducido:")
print(classification_report(y_test, y_pred_reduced))

# Comparar rendimiento entre el modelo original y el modelo reducido
y_pred_full = xgb_model_full.predict(x_test)
print("\nReporte de clasificación para el modelo completo:")
print(classification_report(y_test, y_pred_full))

accuracy_full = accuracy_score(y_test, y_pred_full)
accuracy_reduced = accuracy_score(y_test, y_pred_reduced)

print(f"\nPrecisión del modelo completo: {accuracy_full:.4f}")
print(f"Precisión del modelo reducido: {accuracy_reduced:.4f}")

### 3.3 Calibración Probabilistica [6 puntos]

<center>
<img src='https://media2.giphy.com/media/l2Je4Ku0Cx292KWv6/200w.gif?cid=6c09b952y0sihtq9tb6sz8j2023x3zxxp3qx1ocgonkpkblj&ep=v1_gifs_search&rid=200w.gif&ct=g' width=400 />

Para lograr modelos más modulares, se recomienda realizar una calibración del modelo entrenado anteriormente, con el objetivo de obtener salidas que reflejen mayor modularidad.

1. Se solicita que utilice un método de calibración que asegure que las probabilidades generadas incrementen de manera monótona. Una métrica ampliamente utilizada para evaluar la precisión de la calibración de un modelo es el Brier Score. Calcule el Brier Score para el modelo tanto antes como después de la calibración. Esto le permitirá realizar una comparación cuantitativa y determinar si la calibración ha mejorado el rendimiento del modelo. Para más información sobre el Brier Score, puede consultar el siguiente enlace: [Scikit-Learn - Brier Score Loss](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.brier_score_loss.html). [3 puntos]

2. Tras la calibración, examine y comente los resultados obtenidos. A su análisis añada una comparación visual de las ideales versus las salidas del modelo original (sin calibrar) y del modelo calibrado. [3 puntos]

**Respuesta**

In [ ]:
from sklearn.calibration import CalibratedClassifierCV
from sklearn.metrics import brier_score_loss
import numpy as np

import matplotlib.pyplot as plt

# Entrenar el modelo XGBoost con los datos de entrenamiento
xgb_model = XGBClassifier(random_state=42)
xgb_model.fit(x_train, y_train)

# Calcular el Brier Score antes de la calibración
y_prob_before = xgb_model.predict_proba(x_test)[:, 1]
brier_score_before = brier_score_loss(y_test, y_prob_before)

# Calibrar el modelo utilizando CalibratedClassifierCV con método isotónico
calibrated_model = CalibratedClassifierCV(xgb_model, method='isotonic', cv='prefit')
calibrated_model.fit(x_test, y_test)

# Calcular el Brier Score después de la calibración
y_prob_after = calibrated_model.predict_proba(x_test)[:, 1]
brier_score_after = brier_score_loss(y_test, y_prob_after)

# Comparación visual de las probabilidades ideales vs. las salidas del modelo
plt.figure(figsize=(10, 6))
plt.plot([0, 1], [0, 1], 'k--', label='Ideal')
plt.hist(y_prob_before, bins=50, alpha=0.5, label='Antes de calibración', density=True)
plt.hist(y_prob_after, bins=50, alpha=0.5, label='Después de calibración', density=True)
plt.xlabel('Probabilidad predicha')
plt.ylabel('Densidad')
plt.title('Comparación de calibración del modelo')
plt.legend()
plt.show()

# Imprimir los resultados del Brier Score
print(f"Brier Score antes de la calibración: {brier_score_before:.4f}")
print(f"Brier Score después de la calibración: {brier_score_after:.4f}")

Mucho éxito!

<center>
<img src="https://i.pinimg.com/originals/55/3d/42/553d42bea9b10e0662a05aa8726fc7f4.gif" width=300>